### ISSUE TITLE
phoenix.js contineously reconnects to channels after a socket reconnection

### ISSUE THREAD

**Issue by jpbecotte** [phoenixframework/phoenix#1579]
### Environment

* Elixir version (elixir -v): 1.2.2
* Phoenix version (mix deps): 1.1.4
* NodeJS version (node -v): v4.2.1
* NPM version (npm -v): 2.14.7
* Operating system: OS/X 10.11.3

### Expected behavior
1. Start a phoenix server that handles websocket channels
2. Make a connection with a client
3. Kill the server (CTRL-C, abort) Or kill internet connection
4. Restart the server (or reset internet connection)
5. The connection to the channels is re-established and the phoenix.js library should not try to reconnect to channels once it has been successfully reconnected

### Actual behavior

The phoenix.js library tries to rejoin the channels even though they have been rejoined. So, on the server, we keep getting those messages : 

[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined
[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined
[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined
[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined
[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined
[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined
[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined
[info] HelloPhoenix.LobbyChannel received join event with topic "rooms:lobby" but channel already joined

After some researches and code digging, there's seems to have a bug with the Push object that doesn't clear a timeout.

**Comment by chrismccord** [comment]
Can you include your javascript? The client should not be showing this behaviour and it's not something I can recreate on our end. Thanks!

**Comment by jpbecotte** [comment]
Sure! I tried to reproduce the bug with the https://github.com/chrismccord/phoenix_chat_example without success.

```
angular.module('dupathome').service('socket', function($q, $ionicLoading, AuthService, $rootScope, EndPointService, appstate) {
  var socket = null;
  var lobby = null;
  var privateChannel = null;
  var insureDeferred = false;
  var insuring = false;

  var attachSocketHandlers = function (socket) {
    socket.onOpen(function() {
      console.log("Socket opened. Now joining lobby.");

      lobby = socket.channel("rooms:lobby");
      lobby.on("event", function(event) {
        console.log ("Received event on LOBBY!! ", event);
        $ionicLoading.hide();
        $rootScope.$broadcast(event.event_type, event.event_data);
      });
      lobby.join()
        .receive("error", function(data) {
          console.log("error on lobby", data);
          socket.disconnect();
          reject();
        })
        .receive("ok", function(lobbyData) {
          console.log("lobby joined");
          privateChannel = socket.channel("private:" + AuthService.userData().user_id);
          privateChannel.on("event", function(event) {
            if (event.event_type == 'set_state') {
              console.log ("Received a new state!!!! ", event.event_data.state);
              appstate.go(event.event_data.state, event.event_data.flash_message);
            } else if (event.event_type == 'group_event' && event.event_data.text == 'no_hand') {
              alert("We cannot find a hand. Please come back later!");
            } else {
              console.log ("Received event on private channel : ", event);
              $rootScope.$broadcast(event.event_type, event.event_data);
            }
          });
          privateChannel.join()
            .receive("error", function(data) {
              console.log("error joining private channel", data);
            })
            .receive("ok", function(data) {
              $ionicLoading.hide();

              console.log("private channel joined", data);
              resolve(lobbyData.user_state);
              appstate.go(lobbyData.user_state);
            })
        })
        //.after(10000, function() {console.log("Connection interruption");})
    });

    socket.onError(function(ev) {console.log("SOCKET ERROR", ev);});
    socket.onClose(function(ev) {console.log("SOCKET CLOSE", ev);});
  };

  var attachChannelHandlers = function (chan) {
    chan.onError(function(e) {console.log("Something went wrong", e)});
    chan.onClose(function(e) {console.log("Channel closed", e)});
  };

  var insure = function () {
    if (insuring) return insureDeferred.promise;

    insuring = true;

    insureDeferred = $q.defer();

    if (AuthService.hasToken()) {
      var userData = AuthService.userData() ;
      socket = new Phoenix.Socket(EndPointService.getSocketUrl(), {params: {token: userData.socket_token}});
      attachSocketHandlers(socket);
      socket.connect();
    } else {
      setTimeout(function() {
        console.log("No token found!");
        reject();
      }, 10);
    }

    return insureDeferred.promise;
  }

  var resolve = function(data) {
    insuring = false;
    insureDeferred.resolve(data);
  }

  var reject = function(data) {
    insuring = false;
    insureDeferred.reject(data);
  }

  var isOK = function() {
    return ! insuring && lobby;
  }

  var reset = function() {
    lobby.leave();
    socket.disconnect();
    return insure();
  }

  var push = function (where, term, data, doNotShowSpinner) {
    var deferred = $q.defer();

    if (! doNotShowSpinner) $ionicLoading.show();
    where.push(term, data)
      .receive("ok", function(reply) {
        console.log('replied from '+term+': ', reply);
        if (! doNotShowSpinner) $ionicLoading.hide();
        deferred.resolve(reply);
      })
      .receive("error", function(reply) {
        $ionicLoading.hide();
        console.log('ERROR reply from '+term+': ', reply);
        deferred.reject(reply);
      });

    return deferred.promise;
  }

  return {
    insure: insure,
    reset: reset,
    isOK: isOK,
    pushLobby: function(term, data) {
      return push(lobby, term, data, false);
    },
    pushPrivate: function(term, data) {
      return push(privateChannel, term, data);
    }
  };

})
```

**Comment by jpbecotte** [comment]
I must say that the insure function handles everything. It is called when the app bootstraps if there's a valid socket token and when the login is made thru the api.

**Comment by allyraza** [comment]
I would like to see where you calling that insure function and other client code if possible, I don't think its phoenix client IMO its your client that's making circular calls to that insure function which results in multiple join calls to that channel

**Comment by jerel** [comment]
I ran into something similar a while back (this was in a C# client) and it happened because I joined additional channels from inside the lobby channel's `receive('ok')` callback like you're doing. It seemed that the lobby channel got another `ok` event when the socket reconnected after a server restart but calling `join` again on the other channels would error.

I worked around it with a property in the `receive` callback that kept subsequent calls from calling join but it may be a server side inconsistency.

**Comment by chrismccord** [comment]
As @jerel said, you are seeing this behaviour because you are creating and joining the channel inside the `socket.onOpen` callback, which is fired every time the socket opens. Instead, you should initialize the channel a single time.

**Comment by jpbecotte** [comment]
@jerel and @chrismccord thanks I'll try that! @allyrazano: I made sure it's called only once.

**Comment by jpbecotte** [comment]
I confirm that behaviour. Thanks for your help @jerel and @chrismccord.

**Comment by brajendrak00068** [comment]
opra
